mountpoint-s3-client/src/s3_crt_client/put_object.rs

```diff
@@ -17,117 +16,114 @@ impl S3CrtClient {
+        let span = request_span!(self.inner, "put_object");
         span.in_scope(|| debug!(?bucket, ?key, ?params, "create request"));
-        Self {
-            client,
-            span,
-            bucket,
-            key,
-            params,
-            inner: None,
-        }
-    }
-
-    /// Make a PutObject meta-request.
-    /// To upload an empty object, set `is_empty` to `true` and drop the
-    /// returned [AsyncStreamWriter] without completing it.
-    fn make_request(&self, is_empty: bool) -> Result<S3PutRequest, S3RequestError> {
         let mut message = self
-            .client
-            .new_request_template("PUT", &self.bucket)
+            .inner
+            .new_request_template("PUT", bucket)
             .map_err(S3RequestError::construction_failure)?;
 
-        let key = format!("/{}", self.key);
+        let key = format!("/{}", key);
         message
             .set_request_path(&key)
             .map_err(S3RequestError::construction_failure)?;
 
-        if self.params.trailing_checksums {
+        if params.trailing_checksums {
             let checksum_config = ChecksumConfig::trailing_crc32c();
             message.set_checksum_config(Some(checksum_config));
         }
 
-        let (body_async_stream, writer) = async_stream::new_stream(&self.client.allocator);
+        let (body_async_stream, writer) = async_stream::new_stream(&self.inner.allocator);
         message.set_body_stream(Some(body_async_stream));
 
-        if is_empty {
-            message
-                .add_header(&Header::new("Content-Length", 0.to_string()))
-                .map_err(S3RequestError::construction_failure)?;
-        }
+        span.in_scope(|| debug!(?bucket, ?key, ?params, "make request"));
+
+        let review_callback = ReviewCallbackBox::default();
+        let callback = review_callback.clone();
 
-        self.span
-            .in_scope(|| debug!(bucket=?self.bucket, ?key, params=?self.params, "make request"));
-        let body =
-            self.client
-                .make_simple_http_request(message, MetaRequestType::PutObject, self.span.clone(), |result| {
-                    ObjectClientError::ClientError(S3RequestError::ResponseError(result))
-                })?;
-        Ok((body, writer))
+        let mut options = S3CrtClientInner::new_meta_request_options(message, MetaRequestType::PutObject);
+        options.on_upload_review(move |review| callback.invoke(review));
+        let body = self
+            .inner
+            .make_simple_http_request_from_options(options, span, |result| {
+                ObjectClientError::ClientError(S3RequestError::ResponseError(result))
+            })?;
+
+        Ok(S3PutObjectRequest {
+            body,
+            writer,
+            review_callback,
+        })
     }
 }
 
+type ReviewCallback = dyn FnOnce(UploadReview) -> bool + Send;
+
+#[derive(Clone, Default)]
+struct ReviewCallbackBox {
```

Comment:
Probably deserves a brief comment about why this exists (because the callback is provided at completion time, but its invocation is async).